File: C++/source/teraterm/ttwsk.c
/* Tera Term
 Copyright(C) 1994-1998 T. Teranishi
 All rights reserved. */

/* TERATERM.EXE, Winsock interface */

#include "teraterm.h"
#include "ttwsk.h"

static HANDLE HWinsock = NULL;

Tclosesocket Pclosesocket;
Tconnect Pconnect;
Thtonl Phtonl;
Thtons Phtons;
Tinet_addr Pinet_addr;
Tioctlsocket Pioctlsocket;
Trecv Precv;
Tselect Pselect;
Tsend Psend;
Tsetsockopt Psetsockopt;
Tsocket Psocket;
// Tgethostbyname Pgethostbyname;
TWSAAsyncSelect PWSAAsyncSelect;
TWSAAsyncGetHostByName PWSAAsyncGetHostByName;
TWSACancelAsyncRequest PWSACancelAsyncRequest;
TWSAGetLastError PWSAGetLastError;
TWSAStartup PWSAStartup;
TWSACleanup PWSACleanup;

void CheckWinsock()
{
  WORD wVersionRequired;
  WSADATA WSData;

#ifdef TERATERM32
  if (HWinsock == NULL) return;
#else
  if (HWinsock < HINSTANCE_ERROR) return;
#endif
  wVersionRequired = 1*256+1;
  if ((PWSAStartup(wVersionRequired, &WSData) != 0) ||
      (LOBYTE(WSData.wVersion) != 1) ||
      (HIBYTE(WSData.wVersion) != 1))
  {
    PWSACleanup();
    FreeLibrary(HWinsock);
    HWinsock = NULL;
  }
}

#define IdCLOSESOCKET	  3
#define IdCONNECT	  4
#define IdHTONL 	  8
#define IdHTONS 	  9
#define IdINET_ADDR	  10
#define IdIOCTLSOCKET	  12
#define IdRECV		  16
#define IdSELECT	  18
#define IdSEND		  19
#define IdSETSOCKOPT	  21
#define IdSOCKET	  23
// #define IdGETHOSTBYNAME   52
#define IdWSAASYNCSELECT  101
#define IdWSAASYNCGETHOSTBYNAME 103
#define IdWSACANCELASYNCREQUEST 108
#define IdWSAGETLASTERROR 111
#define IdWSASTARTUP	  115
#define IdWSACLEANUP	  116

BOOL LoadWinsock()
{
  BOOL Err;

#ifdef TERATERM32
  if (HWinsock == NULL)
  {
    HWinsock = LoadLibrary("WSOCK32.DLL");
    if (HWinsock == NULL) return FALSE;
#else
  if (HWinsock < HINSTANCE_ERROR)
  {
    HWinsock = LoadLibrary("WINSOCK.DLL");
    if (HWinsock < HINSTANCE_ERROR) return FALSE;
#endif
    Err = FALSE;

    Pclosesocket = (Tclosesocket)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdCLOSESOCKET));
    if (Pclosesocket==NULL) Err = TRUE;

    Pconnect = (Tconnect)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdCONNECT));
    if (Pconnect==NULL) Err = TRUE;

    Phtonl = (Thtonl)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdHTONL));
    if (Phtonl==NULL) Err = TRUE;

    Phtons = (Thtons)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdHTONS));
    if (Phtons==NULL) Err = TRUE;

    Pinet_addr = (Tinet_addr)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdINET_ADDR));
    if (Pinet_addr==NULL) Err = TRUE;

    Pioctlsocket = (Tioctlsocket)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdIOCTLSOCKET));
    if (Pioctlsocket==NULL) Err = TRUE;

    Precv = (Trecv)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdRECV));
    if (Precv==NULL) Err = TRUE;

    Pselect = (Tselect)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdSELECT));
    if (Pselect==NULL) Err = TRUE;

    Psend = (Tsend)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdSEND));
    if (Psend==NULL) Err = TRUE;

    Psetsockopt = (Tsetsockopt)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdSETSOCKOPT));
    if (Psetsockopt==NULL) Err = TRUE;

    Psocket = (Tsocket)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdSOCKET));
    if (Psocket==NULL) Err = TRUE;

//    Pgethostbyname = (Tgethostbyname)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdGETHOSTBYNAME));
//    if (Pgethostbyname==NULL) Err = TRUE;

    PWSAAsyncSelect = (TWSAAsyncSelect)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdWSAASYNCSELECT));
    if (PWSAAsyncSelect==NULL) Err = TRUE;

    PWSAAsyncGetHostByName =
      (TWSAAsyncGetHostByName)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdWSAASYNCGETHOSTBYNAME));
    if (PWSAAsyncSelect==NULL) Err = TRUE;

    PWSACancelAsyncRequest = (TWSACancelAsyncRequest)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdWSACANCELASYNCREQUEST));
    if (PWSACancelAsyncRequest==NULL) Err = TRUE;

    PWSAGetLastError = (TWSAGetLastError)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdWSAGETLASTERROR));
    if (PWSAGetLastError==NULL) Err = TRUE;

    PWSAStartup = (TWSAStartup)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdWSASTARTUP));
    if (PWSAStartup==NULL) Err = TRUE;

    PWSACleanup = (TWSACleanup)GetProcAddress(HWinsock, MAKEINTRESOURCE(IdWSACLEANUP));
    if (PWSACleanup==NULL) Err = TRUE;

    if (Err)
    {
      FreeLibrary(HWinsock);
      HWinsock = NULL;
      return FALSE;
    }
  }

  CheckWinsock();

#ifdef TERATERM32
  return (HWinsock != NULL);
#else
  return (HWinsock >= HINSTANCE_ERROR);
#endif
}

void FreeWinsock()
{
  HANDLE HTemp;
#ifndef TERATERM32
  MSG Msg;
#endif

#ifdef TERATERM32
  if (HWinsock == NULL) return;
#else
  if (HWinsock < HINSTANCE_ERROR) return;
#endif
  HTemp = HWinsock;
  HWinsock = NULL;
  PWSACleanup();
#ifdef TERATERM32
  Sleep(50); // for safety
#else
  PeekMessage(&Msg,NULL,0,0,PM_NOREMOVE);
#endif
  FreeLibrary(HTemp);
}

File: Pascal/ttwsk.pas
{Tera Term
 Copyright(C) 1994-1998 T. Teranishi
 All rights reserved.}

{TERATERM.EXE, winsock interface}

unit TTWsk;

interface
{$i teraterm.inc}

uses WinTypes, WinProcs, Types, WskTypes;

var
  closesocket: Tclosesocket;
  connect: Tconnect;
  htonl: Thtonl;
  htons: Thtons;
  inet_addr: Tinet_addr;
  ioctlsocket: Tioctlsocket;
  recv: Trecv;
  select: Tselect;
  send: Tsend;
  setsockopt: Tsetsockopt;
  _socket: T_socket;
  {gethostbyname: Tgethostbyname;}
  WSAAsyncSelect: TWSAAsyncSelect;
  WSAAsyncGetHostByName: TWSAAsyncGetHostByName;
  WSACancelAsyncRequest: TWSACancelAsyncRequest;
  WSAGetLastError: TWSAGetLastError;
  WSAStartup: TWSAStartup;
  WSACleanup: TWSACleanup;

function WSAGetAsyncError(lParam: longint): word;

function LoadWinsock: bool;
procedure FreeWinsock;

implementation

var
  HWinsock: THandle;

const
  IdCLOSESOCKET           = 3;
  IdCONNECT               = 4;
  IdHTONL                 = 8;
  IdHTONS                 = 9;
  IdINET_ADDR             = 10;
  IdIOCTLSOCKET           = 12;
  IdRECV                  = 16;
  IdSELECT                = 18;
  IdSEND                  = 19;
  IdSETSOCKOPT            = 21;
  IdSOCKET                = 23;
  {IdGETHOSTBYNAME         = 52;}
  IdWSAASYNCSELECT        = 101;
  IdWSAASYNCGETHOSTBYNAME = 103;
  IdWSACANCELASYNCREQUEST = 108;
  IdWSAGETLASTERROR       = 111;
  IdWSASTARTUP            = 115;
  IdWSACLEANUP            = 116;

procedure CheckWinsock;
var
  wVersionRequired: word;
  WSData: WSADATA;
begin
{$ifdef TERATERM32}
  if HWinsock=0 then exit;
{$else}
  if HWinsock < HINSTANCE_ERROR then exit;
{$endif}
  wVersionRequired := 1*256+1;
  if (WSAStartup(wVersionRequired, @WSData) <> 0) or
     (LO(WSData.wVersion) <> 1) or
     (HI(WSData.wVersion) <> 1) then
  begin
    WSACleanup;
    FreeLibrary(HWinsock);
    HWinsock := 0;
  end;
end;

function LoadWinsock: bool;
var
  Err: bool;
begin
  LoadWinsock := FALSE;
{$ifdef TERATERM32}
  if HWinsock=0 then
  begin
    HWinsock := LoadLibrary('WSOCK32.DLL');
    if HWinsock=0 then exit;
{$else}
  if HWinsock < HINSTANCE_ERROR then
  begin
    HWinsock := LoadLibrary('WINSOCK.DLL');
    if HWinsock < HINSTANCE_ERROR then exit;
{$endif}
    Err := FALSE;

    @closesocket := GetProcAddress(HWinsock, PChar(IdCLOSESOCKET));
    if @closesocket=nil then Err := TRUE;

    @connect := GetProcAddress(HWinsock, PChar(IdCONNECT));
    if @connect=nil then Err := TRUE;

    @htonl := GetProcAddress(HWinsock, PChar(IdHTONL));
    if @htonl=nil then Err := TRUE;

    @htons := GetProcAddress(HWinsock, PChar(IdHTONS));
    if @htons=nil then Err := TRUE;

    @inet_addr := GetProcAddress(HWinsock, PChar(IdINET_ADDR));
    if @inet_addr=nil then Err := TRUE;

    @ioctlsocket := GetProcAddress(HWinsock, PChar(IdIOCTLSOCKET));
    if @ioctlsocket=nil then Err := TRUE;

    @recv := GetProcAddress(HWinsock, PChar(IdRECV));
    if @recv=nil then Err := TRUE;

    @select := GetProcAddress(HWinsock, PChar(IdSELECT));
    if @select=nil then Err := TRUE;

    @send := GetProcAddress(HWinsock, PChar(IdSEND));
    if @send=nil then Err := TRUE;

    @setsockopt := GetProcAddress(HWinsock, PChar(IdSETSOCKOPT));
    if @setsockopt=nil then Err := TRUE;

    @_socket := GetProcAddress(HWinsock, PChar(IdSOCKET));
    if @_socket=nil then Err := TRUE;

    {@gethostbyname := GetProcAddress(HWinsock, PChar(IdGETHOSTBYNAME));
    if @gethostbyname=nil then Err := TRUE;}

    @WSAAsyncSelect := GetProcAddress(HWinsock, PChar(IdWSAASYNCSELECT));
    if @WSAAsyncSelect=nil then Err := TRUE;

    @WSAAsyncGetHostByName := GetProcAddress(HWinsock, PChar(IdWSAASYNCGETHOSTBYNAME));
    if @WSAAsyncGetHostByName=nil then Err := TRUE;

    @WSACancelAsyncRequest := GetProcAddress(HWinsock, PChar(IdWSACANCELASYNCREQUEST));
    if @WSACancelAsyncRequest=nil then Err := TRUE;

    @WSAGetLastError := GetProcAddress(HWinsock, PChar(IdWSAGETLASTERROR));
    if @WSAGetLastError=nil then Err := TRUE;

    @WSAStartup := GetProcAddress(HWinsock, PChar(IdWSASTARTUP));
    if @WSAStartup=nil then Err := TRUE;

    @WSACleanup := GetProcAddress(HWinsock, PChar(IdWSACLEANUP));
    if @WSACleanup=nil then Err := TRUE;

    if Err then
    begin
      FreeLibrary(HWinsock);
      HWinsock := 0;
      exit;
    end;
  end;

  CheckWinsock;

{$ifdef TERATERM32}
  LoadWinsock := HWinsock<>0;
{$else}
  LoadWinsock := HWinsock >= HINSTANCE_ERROR;
{$endif}
end;

procedure FreeWinsock;
var
  HTemp: THandle;
{$ifndef TERATERM32}
  Msg: TMsg;
{$endif}
begin
{$ifdef TERATERM32}
  if HWinsock=0 then exit;
{$else}
  if HWinsock < HINSTANCE_ERROR then exit;
{$endif}
  HTemp := HWinsock;
  HWinsock := 0;
  WSACleanUp;
{$ifdef TERATERM32}
  Sleep(50); {for safety}
{$else}
  PeekMessage(Msg,0,0,0,PM_NOREMOVE);
{$endif}
  FreeLibrary(HTemp);
end;

function WSAGetAsyncError(lParam: longint): word;
begin
  WSAGetAsyncError := HIWORD(lParam);
end;

begin
  HWinsock := 0;
end.
